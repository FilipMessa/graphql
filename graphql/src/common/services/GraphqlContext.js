// @flow

import DataLoader from 'dataloader';

import IdentityDataloader from '../../identity/dataloaders/Identity';
import createBookingLoader from '../../booking/dataloaders/Booking';
import createAirlineLoader from '../../flight/dataloaders/Airline';
import createRatesLoader from '../dataloaders/Rates';
import BookingsLoader from '../../booking/dataloaders/Bookings';
import LocationSuggestionsLoader from '../../location/dataloaders/LocationSuggestions';
import LocationLoader from '../../location/dataloaders/Location';
import FlightLoader from '../../flight/dataloaders/Flight';
import OptionsStorage from './context/OptionsStorage';
import HotelsAvailability from '../../hotel/dataloaders/HotelsAvailability';
import { type SearchParameters as HotelKey } from '../../hotel/dataloaders/flow/SearchParameters';
import HotelByID from '../../hotel/dataloaders/HotelByID';
import HotelCities from '../../hotel/dataloaders/HotelCities';
import HotelRoomsLoader from '../../hotel/dataloaders/HotelRooms';
import HotelRoomAvailabilityLoader from '../../hotel/dataloaders/HotelRoomAvailability';
import HotelRoomBeddingLoader from '../../hotel/dataloaders/RoomBedding';
import PriceStatsLoader from '../../hotel/dataloaders/PriceStats';

import type { Booking } from '../../booking/Booking';
import type { Airline } from '../../flight/Flight';
import type { HotelType } from '../../hotel/dataloaders/flow/HotelType';
import type { HotelCity } from '../../hotel/types/outputs/HotelCity';

/**
 * FIXME: this is stupid - it's already defined by data-loader itself
 */
export type GraphqlContextType = {|
  // DataLoader<K, V>
  apiToken: ?string,
  dataLoader: {|
    airline: DataLoader<string, ?Airline>,
    booking: DataLoader<number | string, Booking>,
    bookings: BookingsLoader,
    flight: FlightLoader,
    identity: IdentityDataloader,
    location: LocationLoader,
    locationSuggestions: LocationSuggestionsLoader,
    rates: DataLoader<string, ?number>,
    hotel: {
      availabilityByLocation: DataLoader<HotelKey, HotelType[]>,
      availabilityByID: DataLoader<HotelKey, HotelType[]>,
      byID: typeof HotelByID,
      cities: DataLoader<string, HotelCity[]>,
      room: HotelRoomsLoader,
      roomAvailability: HotelRoomAvailabilityLoader,
      roomBedding: typeof HotelRoomBeddingLoader,
      priceStats: DataLoader<
        { searchParams: HotelKey, boundary: 'MAX' | 'MIN' },
        number,
      >,
    },
  |},
  options: OptionsStorage,
  _traceCollector?: Object,
|};

export function createContext(token: ?string): GraphqlContextType {
  const bookings = new BookingsLoader(token);
  const locationSuggestions = new LocationSuggestionsLoader();
  const location = new LocationLoader(locationSuggestions);

  return {
    apiToken: token,
    dataLoader: {
      airline: createAirlineLoader(),
      booking: createBookingLoader(token, bookings),
      bookings: bookings,
      flight: new FlightLoader(location),
      identity: new IdentityDataloader(token),
      location: location,
      locationSuggestions: locationSuggestions,
      rates: createRatesLoader(),
      hotel: {
        availabilityByLocation: HotelsAvailability,
        availabilityByID: HotelsAvailability,
        byID: HotelByID,
        cities: HotelCities,
        room: new HotelRoomsLoader(),
        roomAvailability: new HotelRoomAvailabilityLoader(),
        roomBedding: HotelRoomBeddingLoader,
        priceStats: PriceStatsLoader,
      },
    },
    options: new OptionsStorage(),
  };
}
