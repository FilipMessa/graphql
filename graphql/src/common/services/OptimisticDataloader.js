// @flow

import DataLoader from 'dataloader';

// any because it must be very generic (accepts everything)
type K = any;
type V = any;

/**
 * Original data-loader kills load batch if one of this response values is
 * Error instance. This optimistic data-loader keeps Error value as a valid
 * response. This means that it's possible to return as many responses
 * as possible and convert Errors to the error messages in the response
 * if necessary.
 *
 * @see https://github.com/facebook/dataloader/issues/41
 */
export default class OptimisticDataloader extends DataLoader<K, V> {
  constructor(batchLoadFn: Function) {
    super(batchLoadFn);
  }

  loadMany = (keys: K[]): Promise<V[]> => {
    if (!Array.isArray(keys)) {
      throw new TypeError(
        'The loader.loadMany() function must be called with Array<key> ' +
          `but got: ${keys}.`,
      );
    }

    return Promise.all(
      keys.map(key =>
        this.load(key).catch(error => {
          if (error instanceof TypeError) {
            // original data-loader throws TypeErrors
            throw error;
          }
          return error;
        }),
      ),
    );
  };
}
