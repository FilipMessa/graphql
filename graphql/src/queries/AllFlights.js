// @flow

import compareAsc from 'date-fns/compare_asc';
import { GraphQLNonNull, GraphQLList } from 'graphql';
import dateFns from 'date-fns';
import _ from 'lodash';
import request from '../services/HttpRequest';
import config from '../../config/application';
import GraphQLFlight from '../types/Flight';
import FlightsSearchInput from '../types/FlightsSearchInput';
import FlightsOptionsInput from '../types/FlightsOptionsInput';
import type { FlightType } from '../Entities';
import { fetchFlightsFallback } from './flight/Fallback';
import { sanitizeApiResponse } from './flight/ApiSanitizer';

export default {
  type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(GraphQLFlight))),
  args: {
    search: {
      type: new GraphQLNonNull(FlightsSearchInput),
    },
    options: {
      type: FlightsOptionsInput,
    },
  },
  resolve: async (
    ancestor: mixed,
    args: Object,
  ): Promise<Array<FlightType>> => {
    validateArgs(args);

    let allFlights = await request(
      config.restApiEndpoint.allFlights({
        flyFrom: args.search.from,
        to: args.search.to,
        dateFrom: dateFns.format(new Date(args.search.dateFrom), 'DD/MM/YYYY'),
        dateTo: dateFns.format(new Date(args.search.dateTo), 'DD/MM/YYYY'),
        curr: _.get(args, 'options.currency'),
      }),
    );

    if (allFlights._results === 0) {
      allFlights = await fetchFlightsFallback(args);
    }

    return allFlights.data.map(flight =>
      sanitizeApiResponse(flight, allFlights.currency),
    );
  },
};

function validateArgs(args: Object) {
  // Validate dateFrom starts before dateTo
  const dateFrom = new Date(args.search.dateFrom);
  const dateTo = new Date(args.search.dateTo);
  if (compareAsc(dateFrom, dateTo) > 0) {
    throw new Error(`DateFrom should start before dateTo`);
  }
}
