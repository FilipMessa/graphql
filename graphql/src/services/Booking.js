import jsonMock from '../../tests/bookings.json';

// const url = `https://booking-api.skypicker.com/api/v0.1/users/self/bookings?token=${token}`;

export default () => jsonMock; // FIXME: replace with API call
