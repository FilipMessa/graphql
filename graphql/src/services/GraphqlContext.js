// @flow

import DataLoader from 'dataloader';
import createIdentityLoader from '../dataLoaders/Identity';
import createBookingLoader from '../dataLoaders/Booking';
import createAirlineLoader from '../dataLoaders/Airline';
import BookingsLoader from '../dataLoaders/Bookings';
import type { IdentityType, BookingType, AirlineType } from '../Entities';

export type GraphqlContextType = {
  apiToken: ?string,
  dataLoader: {
    booking: DataLoader<number | string, BookingType>,
    bookings: BookingsLoader,
    identity: DataLoader<string, IdentityType>,
    airline: DataLoader<string, ?AirlineType>,
  },
};

export function createContext(token: ?string): GraphqlContextType {
  const bookings = new BookingsLoader(token);
  return {
    apiToken: token,
    dataLoader: {
      booking: createBookingLoader(token, bookings),
      bookings: bookings,
      identity: createIdentityLoader(token),
      airline: createAirlineLoader(),
    },
  };
}
