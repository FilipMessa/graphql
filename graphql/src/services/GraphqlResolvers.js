// @flow

import _ from 'lodash';
import differenceInMinutes from 'date-fns/difference_in_minutes';
import type { ArrivalType, DepartureType } from '../Entities';

export function flightDurationInMinutes(
  departure: DepartureType,
  arrival: ArrivalType,
): ?number {
  const departureTime = _.get(departure, 'when.utc');
  const arrivalTime = _.get(arrival, 'when.utc');
  return dateDurationInMinutes(departureTime, arrivalTime);
}

export function dateDurationInMinutes(departure: ?Date, arrival: ?Date) {
  if (!departure || !arrival || departure > arrival) {
    return null;
  }
  return differenceInMinutes(arrival, departure);
}
