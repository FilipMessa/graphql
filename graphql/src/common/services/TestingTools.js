// @flow

import { graphql as graphqlNodeLibrary } from 'graphql';
import schema from '../../Schema';
import { createContext } from './GraphqlContext';

/**
 * TestingTools must be imported FIRST otherwise mock of HttpRequest won't work.
 */
jest.mock('../services/HttpRequest');
const mockedHttpRequest = require('./HttpRequest');

class MockResponse {
  absoluteUrl: string;
  httpMethod: string;

  constructor(httpMethod: 'GET' | 'POST', absoluteUrl: string) {
    this.httpMethod = httpMethod;
    this.absoluteUrl = absoluteUrl;
  }

  replyWithData(response: Array<Object> | Object) {
    // $FlowAllowMockMagic
    mockedHttpRequest.__setMockData(
      this.httpMethod,
      this.absoluteUrl,
      response,
    );
  }
}

export class RestApiMock {
  static onGet(absoluteUrl: string) {
    return new MockResponse('GET', absoluteUrl);
  }

  static onPost(absoluteUrl: string) {
    return new MockResponse('POST', absoluteUrl);
  }
}

export const graphql = async (
  query: string,
  variables: ?Object,
): Promise<Object> =>
  graphqlNodeLibrary(
    schema,
    query,
    null,
    createContext('test_token'),
    variables,
  );

export const evaluateResolver = (field: mixed, testValue: mixed) => {
  // $FlowAllowNextLineInThisTest (fields are possibly undefined but we assume it's fine)
  return field.resolve(testValue);
};
