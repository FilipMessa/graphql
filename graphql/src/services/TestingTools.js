// @flow

import { graphql } from 'graphql';
import schema from '../Schema';
import { createContext } from '../services/GraphqlContext';

jest.mock('../services/HttpRequest');

export const executeQuery = async (
  query: string,
  variables: ?Object,
): Promise<Object> =>
  graphql(schema, query, null, createContext('test_token'), variables);
