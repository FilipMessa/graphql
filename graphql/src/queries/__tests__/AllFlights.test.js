// @flow

import { executeQuery } from '../../services/TestingTools';
import AllFlights from '../AllFlights';

describe('all flights query', () => {
  it('should be non-null list of non-null Flight types', () => {
    expect(AllFlights.type.toString()).toBe('[Flight!]!');
  });

  it('should return array of flights', async () => {
    const allFlightsSearchQuery = `
    query ($input: FlightsSearchInput!) {
      allFlights(search: $input) {
        arrival {
          airport { city { name }, code }
          time, localTime
        }
        departure {
          airport { city { name }, code }
          time, localTime
        }
        airlines {
          name
          code
          logoUrl
          isLowCost
        }
      }
    }`;
    const variables = {
      input: {
        from: 'PRG',
        to: 'MEX',
        dateFrom: '2017-08-08',
        dateTo: '2017-09-08',
      },
    };
    expect(
      await executeQuery(allFlightsSearchQuery, variables),
    ).toMatchSnapshot();
  });

  it('should return error if invalid date format is passed', async () => {
    const allFlightsSearchQuery = `
    query ($input: FlightsSearchInput!) {
      allFlights(search: $input) {
        arrival {
          time
        }
      }
    }`;
    const variables = {
      input: {
        from: 'PRG',
        to: 'MEX',
        dateFrom: '08/08/2017',
        dateTo: '2017-09-08',
      },
    };
    expect(
      await executeQuery(allFlightsSearchQuery, variables),
    ).toMatchSnapshot();
  });

  it('should return error if dateFrom is after dateTo', async () => {
    const allFlightsSearchQuery = `
    query ($input: FlightsSearchInput!) {
      allFlights(search: $input) {
        arrival {
          time
        }
      }
    }`;
    const variables = {
      input: {
        from: 'PRG',
        to: 'MEX',
        dateFrom: '2018-08-08',
        dateTo: '2017-09-08',
      },
    };
    expect(
      await executeQuery(allFlightsSearchQuery, variables),
    ).toMatchSnapshot();
  });
});
