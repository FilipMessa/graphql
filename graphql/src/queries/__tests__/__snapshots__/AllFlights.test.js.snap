// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all flights query should return array of flights 1`] = `
Object {
  "data": Object {
    "allFlights": Array [
      Object {
        "arrival": Object {
          "airport": Object {
            "city": "Mexico City",
            "code": "MEX",
          },
          "localTime": "2017-09-08T12:45:00.000Z",
          "time": "2017-09-08T17:45:00.000Z",
        },
        "departure": Object {
          "airport": Object {
            "city": "Prague",
            "code": "PRG",
          },
          "localTime": "2017-09-06T11:40:00.000Z",
          "time": "2017-09-06T09:40:00.000Z",
        },
      },
      Object {
        "arrival": Object {
          "airport": Object {
            "city": "Mexico City",
            "code": "MEX",
          },
          "localTime": "2017-09-08T05:30:00.000Z",
          "time": "2017-09-08T10:30:00.000Z",
        },
        "departure": Object {
          "airport": Object {
            "city": "Prague",
            "code": "PRG",
          },
          "localTime": "2017-09-06T11:40:00.000Z",
          "time": "2017-09-06T09:40:00.000Z",
        },
      },
      Object {
        "arrival": Object {
          "airport": Object {
            "city": "Mexico City",
            "code": "MEX",
          },
          "localTime": "2017-09-08T12:45:00.000Z",
          "time": "2017-09-08T17:45:00.000Z",
        },
        "departure": Object {
          "airport": Object {
            "city": "Prague",
            "code": "PRG",
          },
          "localTime": "2017-09-06T14:05:00.000Z",
          "time": "2017-09-06T12:05:00.000Z",
        },
      },
    ],
  },
}
`;

exports[`all flights query should return error if invalid date format is passed 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Variable "$input" got invalid value {"from":"PRG","to":"MEX","dateFrom":"08/08/2017","dateTo":"2017-09-08"}.
In field "dateFrom": Expected type "Date", found "08/08/2017": Date cannot represent an invalid date-string 08/08/2017.],
  ],
}
`;
